Migrate API defaulting to a centralized location

yujuhong opened the pull request:
This PR is based on #2587 to fix issue #1502. Most API defaulting in validation.go and conversion.go have been migrated to defaults.go, where the defaulting functions are added to the decoder. The defaulting functions are tied to the specific API version, and will be called during decoding.

thockin committed:
Sketch: a third take on defaulting values

yujuhong committed:
Factor out API defaulting from validation logic

Currently, the validation logic validates fields in an object and supply default
values wherever applies. This change factors out defaulting to a set of
defaulting callback functions for decoding (see #1502 for more discussion).

 * This change is based on pull request 2587.

 * Most defaulting has been migrated to defaults.go where the defaulting
   functions are added.

 * validation_test.go and converter_test.go have been adapted to not testing the
   default values.

 * Fixed all tests with that create invalid objects with the absence of
   defaulting logic.

alex-mohr commented:
@bgrant0607 Brian, you were pretty active in the referenced #1502 discussion -- do you have time to review?

yujuhong commented:
@thockin, this PR will need a rebase to fix all recently modified tests once it is reviewed.

bgrant0607 commented:
@alex-mohr Sorry, I had a massive backlog yesterday. I'll look at this today.

@smarterclayton May also want a look.

bgrant0607 commented on pkg/api/defaults.go:
```diff
@@ -0,0 +1,65 @@
+			if util.AllPtrFieldsNil(obj) {
+				obj.Always = &RestartPolicyAlways{}
+			}
+		},
```
DNSPolicy default should be set here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta1/types.go#L65
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/validation/validation.go#L378

bgrant0607 replied:
These comments apply to all the versioned instances of defaults.go.

bgrant0607 commented on pkg/api/defaults.go:
```diff
@@ -0,0 +1,65 @@
+			if obj.Spec.Protocol == "" {
+				obj.Spec.Protocol = ProtocolTCP
+			}
+		},
```
SessionAffinity:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta1/types.go#L516
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/validation/validation.go#L489

bgrant0607 commented on pkg/conversion/converter.go:
```diff
@@ -210,14 +214,18 @@ func (s *scope) error(message string, args ...interface{}) error {
 // Example:
-// c.Register(func(in *Pod, out *v1beta1.Pod, s Scope) error { ... return nil })
-func (c *Converter) Register(conversionFunc interface{}) error {
+// c.RegisteConversionFuncr(
```
Typo

bgrant0607 commented on pkg/conversion/converter.go:
```diff
@@ -266,6 +274,33 @@ func (c *Converter) SetStructFieldCopy(srcFieldType interface{}, srcFieldName st
+// defaultingFunc must take one parameters: a pointer to the input type.
+//
+// Example:
+// c.RegisteDefaultingFuncr(
```
Typo

bgrant0607 commented on pkg/conversion/converter.go:
```diff
@@ -379,7 +414,18 @@ func (c *Converter) callCustom(sv, dv, custom reflect.Value, scope *scope) error
+	// Apply default values.
+	if fv, ok := c.defaultingFuncs[st]; ok {
+		if c.Debug != nil {
+			c.Debug.Logf("Applying defaults for '%v'", st)
```
Please add a TODO to convert this to glog.

smarterclayton replied:
Part of the reason it's not glog is that in test cases glog is not properly interspersed with "t.Errorf" messages, which makes debugging tables very very difficult.  It's probably better to leave this, since the only real case for this log is test suites.

bgrant0607 replied:
Ok.

smarterclayton commented on pkg/api/v1beta3/defaults.go:
```diff
@@ -0,0 +1,66 @@
+			}
+		},
+		func(obj *Container) {
+			if obj.ImagePullPolicy == "" {
```
nit: We use len(string) == 0 instead of == ""

thockin replied:
We are in no way consistent with this :)

smarterclayton commented:
Good changes in general.

bgrant0607 commented on pkg/conversion/converter.go:
```diff
@@ -266,6 +274,33 @@ func (c *Converter) SetStructFieldCopy(srcFieldType interface{}, srcFieldName st
+//         func(in *v1beta1.Pod) {
+//                 // defaulting logic...
+//          })
+func (c *Converter) RegisterDefaultingFunc(defaultingFunc interface{}) error {
```
We need a test explicitly for the new defaulting code. The existing tests were implicitly relying on the ad hoc default approaches that were previously used. Now they're not.

bgrant0607 commented on pkg/api/defaults.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+func init() {
+	Scheme.AddDefaultingFuncs(
```
In the spirit of #3914, we shouldn't be adding any defaulting functions here. Let's just eliminate this file.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -38,8 +38,7 @@ func validateVolumes(volumes []api.Volume) (util.StringSet, errs.ValidationError
 	allNames := util.StringSet{}
-	for i := range volumes {
-		vol := &volumes[i] // so we can set default values
+	for i, vol := range volumes {
```
This is a good way to prove that we are not depending on mutating changes, but it is pessimal for performance - it will make a copy  of every object.  I think we should stick with pointers.

smarterclayton replied:
Our encoding / decoding / conversion is an order of magnitude worse (we do 2000 allocations per decode, and actually encode 3 times).  A few copies we can probably let slide...


> On Jan 29, 2015, at 7:31 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/validation/validation.go:
> 
> > @@ -38,8 +38,7 @@ func validateVolumes(volumes []api.Volume) (util.StringSet, errs.ValidationError
> >  	allErrs := errs.ValidationErrorList{}
> >  
> >  	allNames := util.StringSet{}
> > -	for i := range volumes {
> > -		vol := &volumes[i] // so we can set default values
> > +	for i, vol := range volumes {
> This is a good way to prove that we are not depending on mutating changes, but it is pessimal for performance - it will make a copy of every object. I think we should stick with pointers.
> 
> —
> Reply to this email directly or view it on GitHub.
>

thockin replied:
Why pessimize when we don't really have to? I leave it to Yu-Ju - if it is a pain to convert back, skip it for now..

thockin commented:
Overall I am good with this.  I do dislike the need to sprinkle default values around the test codebase - if we got rid of the internal defaulting logic, do es that need go away?

bgrant0607 commented:
@thockin With this change, validation depends on defaults being set. As we discussed, the "right" thing to do would be to create all of these test objects by decoding versioned objects. As a convenience and optimization, we could create a library of creators of simple objects for common scenarios -- internally we call these "test fixtures". That would be a pretty big change, though. I don't think we should block this PR on it.

thockin commented:
Agree no need to block on the constructors - this is strictly an improvement.

yujuhong commented:
Addressed the comments in the new patch.
  * Added some (not complete) tests for defaulting (defaults_test.go). 
  * Modified DecodeInto() to always convert even if the versions are the same (e.g. v1beta1 -> v1beta1). This is desirable because we want the side-effect of applying the defaults even when converting a versioned json/yaml to the same version of object. This facilitates the testing in defaults_test.go
  * Add a DeepDerivative helper function in deep_equal.go. This function compares two objects, a1, and a2, but ignore the unset/empty fields in a1. This is useful because in many cases we compare the expected result (a1) to the actual result (a2), while all we care about are the *set* fields in a1. This is particular useful for avoiding comparing the default values in many tests.

This PR has grown bigger and bigger every time I had to rebase. Hopefully we can get it in before another major rebase is required...

bgrant0607 commented:
Thanks! LGTM. I'm happy to start with this, then improve it. Did you run e2e?

thockin commented:
I am concerned about yet another deep-something comparator, but we can revisit after we lock this in :)

yujuhong commented:
@thockin , I understand the concern, and we should definitely iterate on that later. My guess is that many tests are actually redundant, as they test the same logic repeatedly. A major cleanup might make these tests more manageable, and the comparator may no longer required(?).

yujuhong commented:
One of the test suite in e2e failed:
```
2015/02/03 16:31:57 Passed tests: certs.sh[1/1] guestbook.sh[1/1] liveness.sh[1/1] monitoring.sh[1/1] pd.sh[1/1] services.sh[1/1] update.sh[1/1]
2015/02/03 16:31:57 Flaky tests:
2015/02/03 16:31:57 Failed tests: goe2e.sh[0/1]
2015/02/03 16:31:57 1 test(s) failed.
```
`goe2e.sh` failed due to errors such as `x509: certificate signed by unknown authority`
I doubt my change caused the failure, but I am not sure how to resolve this yet.

thockin commented:
the x509 error is the bane of my job.  Nuke your cluster and start over.

On Tue, Feb 3, 2015 at 4:40 PM, Yu-Ju Hong <notifications@github.com> wrote:

> One of the test suite in e2e failed:
>
> 2015/02/03 16:31:57 Passed tests: certs.sh[1/1] guestbook.sh[1/1] liveness.sh[1/1] monitoring.sh[1/1] pd.sh[1/1] services.sh[1/1] update.sh[1/1]
> 2015/02/03 16:31:57 Flaky tests:
> 2015/02/03 16:31:57 Failed tests: goe2e.sh[0/1]
> 2015/02/03 16:31:57 1 test(s) failed.
>
> goe2e.sh failed due to errors such as x509: certificate signed by unknown
> authority
> I doubt my change caused the failure, but I am not sure how to resolve
> this yet.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3854#issuecomment-72767986>
> .
>

yujuhong commented:
e2e finally passed. The PR should be ready to merge.

bgrant0607 commented:
Great! Re-running travis.

bgrant0607 commented:
Travis passed. Shippable is 2 hours behind as usual. Merging.
